File: app/src/main/java/io/oldering/tvfoot/red/data/entity/search/Where.java
package io.oldering.tvfoot.red.data.entity.search;

import com.google.auto.value.AutoValue;

@AutoValue public abstract class Where {
  public static Builder builder() {
    return new AutoValue_Where.Builder().deleted(Deleted.builder().build());
  }

  @AutoValue.Builder public static abstract class Builder {
    public abstract Builder deleted(Deleted deleted);

    public abstract Where build();
  }

  abstract Deleted deleted();

  @Override public String toString() {
    return "{" + "\"deleted\":" + deleted() + "}";
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/data/entity/search/Where.kt
package com.benoitquenaudon.tvfoot.red.app.data.entity.search

import com.benoitquenaudon.tvfoot.red.util.TeamCode
import com.squareup.moshi.formatIso8601
import java.util.Date
import java.util.concurrent.TimeUnit

data class Where(
    private val startAt: Long,
    private val onlyBroadcasted: Boolean = false,
    private val teams: List<TeamCondition> = emptyList()
) {
  override fun toString(): String {
    val buffer = mutableListOf<String>()

    buffer.add(""""start-at":{"gte":"${Date(startAt).formatIso8601()}"}""")
    buffer.add(""""deleted":{"neq":1}""")
    if (onlyBroadcasted) buffer.add(""""broadcasters":{"gt":[]}""")
    if (teams.isNotEmpty()) buffer.add(""""or":[${teams.joinToString(",")}]""")

    return buffer.joinToString(prefix = "{", separator = ",", postfix = "}")
  }

  data class TeamCondition(
      private val code: TeamCode
  ) {
    override fun toString() =
        """{"home-team.code":"$code"},{"away-team.code":"$code"}"""
  }
}
